pdns/dnsdist.cc

```diff
@@ -588,15 +592,15 @@ try {
         }
 
         if (ids->cs && !ids->cs->muted) {
-          if (ids->du) {
+          if (du) {
 #ifdef HAVE_DNS_OVER_HTTPS
             // DoH query
-            ids->du->query = std::string(response, responseLen);
-            if (send(ids->du->rsock, &ids->du, sizeof(ids->du), 0) != sizeof(ids->du)) {
-              delete ids->du;
+            du->response = std::string(response, responseLen);
+            if (send(du->rsock, &du, sizeof(du), 0) != sizeof(du)) {
+              delete du;
```

Comment:
Hmm, is `ids->du` stll the same pointer? That would now be dangling...

Reply:
Unless I'm mistaken, the only case where `du` is not a `NULL` pointer is when `ids->origFD` was still equal to `oldFD` after we atomically updated it, meaning that the `oldFD == origFD` condition on line 566 was true and therefore `ids->du` was set to `nullptr`.

Reply:
A comment and/or an assert on that `ids->du == nullptr` condition would be nice

Reply:
I pushed a bunch of comments about our `IDState`s object and in particular its `origFD` member since it's probably the hardest part of dnsdist to grasp. We can't add an assertion because it would be racy, the state might have already been reused by another thread at that point.

Reply:
👍